src/tools/quemap/lightmap.c

```diff
@@ -903,6 +935,87 @@ void IndirectLightingImpact(const cm_trace_t *trace, const vec3_t color) {
 	}
 }
 
+/**
+* @brief
+*/
+void BuildIndirectLights(void) {
+	int32_t i, face_num;
+
+	num_lights = 0;
+	memset(lights, 0, sizeof(lights));
+
+	for (face_num = 0; face_num < bsp_file.num_faces; face_num++) {
+		const bsp_face_t *face = &bsp_file.faces[face_num];
+		const bsp_texinfo_t *texinfo = &bsp_file.texinfo[face->texinfo];
+		const bsp_plane_t *plane = &bsp_file.planes[face->plane_num];
+
+		if (texinfo->flags & (SURF_SKY | SURF_WARP | SURF_LIGHT | SURF_HINT | SURF_NO_DRAW)) {
+			continue; // we have no light to reflect
+		}
+
+		const face_lighting_t *fl = &face_lighting[face_num];
+
+		vec3_t normal;
+
+		if (face->side) {
+			VectorNegate(plane->normal, normal);
+		}
+		else {
+			VectorCopy(plane->normal, normal);
+		}
+
+		const vec_t *center = face_extents[face_num].center;
+
+		for (i = 0; i < fl->num_samples; i++) {
+			byte pvs[(MAX_BSP_LEAFS + 7) / 8];
+
+			vec3_t direct, indirect, color;
+			VectorCopy(fl->direct + i * 3, direct);
+			VectorCopy(fl->indirect + i * 3, indirect);
+			VectorAdd(direct, indirect, color);
+
+			if (VectorCompare(color, vec3_origin)) {
+				continue;
+			}
+
+			vec3_t point, position;
+
+			VectorCopy(fl->origins + i * 3, point);
+
+			if (texinfo->flags & SURF_PHONG) {
+				SampleNormal(face, position, normal);
+			}
+
+			vec3_t offset;
+			VectorScale(normal, 1.0, offset);
+
+			VectorAdd(point, offset, point);
+
+			if (!NudgeSamplePosition(point, normal, center, position, pvs)) {
+				continue;
+			}			
+
+			num_lights++;
+			light_t *light = Mem_TagMalloc(sizeof(*light), MEM_TAG_LIGHT);
+
+			VectorCopy(position, light->origin);
+
+			const bsp_leaf_t *leaf = &bsp_file.leafs[Light_PointLeafnum(light->origin)];
+			const int16_t cluster = leaf->cluster;
+			light->next = lights[cluster];
+			lights[cluster] = light;
+
+			light->type = LIGHT_SPOT;
+			light->stopdot = 1.0;
+
+			VectorCopy(normal, light->normal);
+
+			VectorScale(color, 1.0 / 255.0 / 256.0 / (float)current_bounce, light->color);
+			light->intensity = 256;
```

Comment:
So they all uniformly bounce light 256 units away, but the light is scaled by the direct light color?

Shouldn't we factor the material hardness in here? We have easy access to it..

Reply:
I'm not suggesting we get all fancy with it. Just multiply the reflected light by the material hardness 🤓 That's literally what it does in the shader.

Reply:
Hardness only scales the specular light, not the diffuse :)

Reply:
Touché.